`std.zig.c_translation.cast` fails to cast function body to function pointer

### Zig Version

0.14.0

### Steps to Reproduce and Observed Behavior

Original report: https://github.com/castholm/SDL/issues/11

This is a translate-c bug. Reduced repro (from `@cImport`ed code):

```zig
const SDL_FunctionPointer = ?*const fn () callconv(.c) void;

fn SDL_BeginThreadFunction(_: ?*anyopaque, _: c_uint, _: ?*const fn (?*anyopaque) callconv(.c) c_uint, _: ?*anyopaque, _: c_uint, _: [*c]c_uint) callconv(.c) usize {
    return 0;
}

comptime {
    _ = @import("std").zig.c_translation.cast(SDL_FunctionPointer, SDL_BeginThreadFunction);
}
```

```sh
zig build-obj repro.zig -fno-emit-bin -freference-trace
```

```
C:\zig\zig\0.14.0\files\lib\std\zig\c_translation.zig:98:26: error: expected type '*const fn () callconv(.c) void', found '*const fn (?*anyopaque, c_uint, ?*const fn (?*anyopaque) callconv(.c) c_uint, ?*anyopaque, c_uint, [*c]c_uint) callconv(.c) usize'
    return @as(DestType, target);
                         ^~~~~~
C:\zig\zig\0.14.0\files\lib\std\zig\c_translation.zig:98:26: note: pointer type child 'fn (?*anyopaque, c_uint, ?*const fn (?*anyopaque) callconv(.c) c_uint, ?*anyopaque, c_uint, [*c]c_uint) callconv(.c) usize' cannot cast into pointer type child 'fn () callconv(.c) void'
C:\zig\zig\0.14.0\files\lib\std\zig\c_translation.zig:98:26: note: return type 'usize' cannot cast into return type 'void'
C:\zig\zig\0.14.0\files\lib\std\zig\c_translation.zig:16:33: note: called from here
                return castToPtr(DestType, SourceType, target);
                       ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
repro.zig:8:46: note: called from here
    _ = @import("std").zig.c_translation.cast(SDL_FunctionPointer, SDL_BeginThreadFunction);
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
referenced by:
    root: C:\zig\zig\0.14.0\files\lib\std\start.zig:3:22
    comptime: C:\zig\zig\0.14.0\files\lib\std\start.zig:27:9
    start: C:\zig\zig\0.14.0\files\lib\std\std.zig:97:27
    comptime: C:\zig\zig\0.14.0\files\lib\std\std.zig:168:9
```

### Expected Behavior

No compile errors.

Suggested fix:
```diff
 fn castToPtr(comptime DestType: type, comptime SourceType: type, target: anytype) DestType {
     switch (@typeInfo(SourceType)) {
         .int => {
             return @as(DestType, @ptrFromInt(castInt(usize, target)));
         },
         .comptime_int => {
             if (target < 0)
                 return @as(DestType, @ptrFromInt(@as(usize, @bitCast(@as(isize, @intCast(target))))))
             else
                 return @as(DestType, @ptrFromInt(@as(usize, @intCast(target))));
         },
         .pointer => {
             return castPtr(DestType, target);
         },
+        .@"fn" => {
+            return castPtr(DestType, &target);
+        },
         .optional => |target_opt| {
             if (@typeInfo(target_opt.child) == .pointer) {
                 return castPtr(DestType, target);
             }
         },
         else => {},
     }
     return @as(DestType, target);
 }
```

--- a/lib/std/zig/c_translation.zig
+++ b/lib/std/zig/c_translation.zig
@@ -88,6 +88,9 @@ fn castToPtr(comptime DestType: type, comptime SourceType: type, target: anytype
         .pointer => {
             return castPtr(DestType, target);
         },
+        .@"fn" => {
+            return castPtr(DestType, &target);
+        },
         .optional => |target_opt| {
             if (@typeInfo(target_opt.child) == .pointer) {
                 return castPtr(DestType, target);
@@ -685,4 +688,15 @@ test "Extended C ABI casting" {
         try testing.expect(@TypeOf(Macros.L_SUFFIX(@as(c_long, math.maxInt(c_long) - 1))) == c_long); // c_long
         try testing.expect(@TypeOf(Macros.L_SUFFIX(math.maxInt(c_long) + 1)) == c_longlong); // comptime_int -> c_longlong
     }
+}
+
+// Function with complex signature for testing the SDL case
+fn complexFunction(_: ?*anyopaque, _: c_uint, _: ?*const fn (?*anyopaque) callconv(.c) c_uint, _: ?*anyopaque, _: c_uint, _: [*c]c_uint) callconv(.c) usize {
+    return 0;
+}
+
+test "function pointer casting" {
+    const SDL_FunctionPointer = ?*const fn () callconv(.c) void;
+    const fn_ptr = cast(SDL_FunctionPointer, complexFunction);
+    try testing.expect(fn_ptr != null);
 }

